File: backend_py/location.py

from flask import jsonify
from db import connect
import boto3
import os
from botocore.exceptions import BotoCoreError, ClientError

# Cargar las variables de entorno
ACCESS_KEY = os.getenv('ACCESSKEY')
SECRET_ACCESS_KEY = os.getenv('SECRETACCESSKEY')
AWS_REGION = os.getenv('AWSREGION')
PLACE_INDEX_NAME = os.getenv('PLACE_INDEX_NAME')

# Configuración del cliente de AWS Location
location_client = boto3.client(
    'location',
    aws_access_key_id=ACCESS_KEY,
    aws_secret_access_key=SECRET_ACCESS_KEY,
    region_name=AWS_REGION
)

# Función para buscar una ubicación por nombre
def buscar_lugar_por_nombre(nombre_lugar):
    params = {
        'IndexName': PLACE_INDEX_NAME,
        'Text': nombre_lugar,
        'MaxResults': 1
    }

    try:
        response = location_client.search_place_index_for_text(**params)
        results = response.get('Results', [])

        if results:
            point = results[0]['Place']['Geometry']['Point']  # [longitud, latitud]
            return jsonify({"coordenadas": point})
        else:
            raise ValueError("No se encontró el lugar.")
    except (BotoCoreError, ClientError) as error:
        raise RuntimeError(f"Error buscando el lugar: {error}")

File: backend_js/routes/location.js
const express = require('express');
const AWS = require('aws-sdk');
require('dotenv').config();

const router = express.Router();

// Configuración de AWS
AWS.config.update({
  accessKeyId: process.env.ACCESSKEY,
  secretAccessKey: process.env.SECRETACCESSKEY,
  region: process.env.AWSREGION
});

// Crear instancia del servicio de localización
const locationService = new AWS.Location();

// Función para buscar ubicación por nombre
const buscarLugarPorNombre = async (nombreLugar) => {
  const params = {
    IndexName: process.env.PLACE_INDEX_NAME,
    Text: nombreLugar,
    MaxResults: 1
  };

  return new Promise((resolve, reject) => {
    locationService.searchPlaceIndexForText(params, (err, data) => {
      if (err) {
        reject("Error buscando el lugar: " + err);
      } else {
        const place = data.Results[0]?.Place;
        if (place) {
          resolve(place.Geometry.Point); // [longitud, latitud]
        } else {
          reject("No se encontró el lugar.");
        }
      }
    });
  });
};

// Ruta para buscar lugar por nombre (GET)
// Ruta para buscar lugar por nombre (GET)
router.get('/buscar-lugar', async (req, res) => {
    const { nombre_lugar } = req.query; // Asegúrate de extraer 'nombre_lugar' de los query parameters
    if (!nombre_lugar) {
      return res.status(400).json({ error: 'El parámetro nombre_lugar es requerido.' });
    }
  
    try {
      const coordenadas = await buscarLugarPorNombre(nombre_lugar);
      res.json({ coordenadas });
    } catch (error) {
      res.status(500).json({ error });
    }
  });
  

// Exportar el router
module.exports = router;
